docs/functionaldatatypes/zset.md

```diff
@@ -0,0 +1,276 @@
+---
+id: zset
+title: "ZSet"
+---
+
+`ZSet` is a version of a set that generalizes the measure of "how many times" each element appears in the set.
+
+Conceptually, we can think of a `ZSet[A, B]` as a `Map[A, B]`, where `A` represents the elements of the set and `B` represents some notion of how many times each element appears.
+
+We can then define a `Set` as a `ZSet` where the measure of how many times each element appears is a `Boolean`.
+
+```scala mdoc
+import zio.prelude.ZSet
+
+type Set[+A] = ZSet[A, Boolean]
+```
+
+That is, in a `Set` the measure of how many times an element appears is just yes or no. Either the element is in the set or it is not.
+
+However, we can also define many other variants of sets by choosing different types to measure how many times an element appears.
+
+For example, by using the `Natural` new type from ZIO Prelude, which describes integers greater than or equal to zero, we can define the type of a `MultiSet` or "bag" in some programming languages.
+
+```scala mdoc:reset
+import zio.prelude.ZSet
+import zio.prelude.newtypes.Natural
+
+type MultiSet[+A] = ZSet[A, Natural]
+```
+
+This ends up being a quite useful data type and can represent anything where the number of times an element appears matters but the order doesn't matter, such as a shopping cart.
+
+By using other measures of how many times an item appears we can describe other variants of sets such as `Int` for sets where items can appear a negative number of times or `Double` for sets where the number of times an item appears can be fractional.
```

Comment:
semi-pedantic: I believe you but this usage is rather novel to me, maybe an external reference to common usage of such idioms (or a quote)? I can imagine a debt of $5.03 being a negative double set of dollars but I am not sure I would want to model that as a "generalized" set.